### PR TITLE
Added create record functionality

### PR DESCRIPTION
**Pull request by AnastasiaShemyakinskaya**
Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

Signed-off-by: Anastasia Shemyakinskaya shem98a@gmail.com

### CODE REVIEW

File: src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java [hunk 1]
```diff
@@ -27,17 +61,31 @@ public KafkaStorageDriver(
...
+    this.linger = netConfig.intVal("linger");
+    this.numPartition = 1;
+    this.replicationFactor = 1;
   }
 
   @Override
   protected final boolean submit(final O op) throws IllegalStateException {
     if (concurrencyThrottle.tryAcquire()) {
       final var opType = op.type();
-      if (opType.equals(OpType.NOOP)) {
```
Comment (paulgrape) [inline src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java]: Why did you remove NOOP check?

File: src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java [hunk 2]
```diff
@@ -12,13 +14,45 @@
...
+  private int batch;
+  private int sndBuf;
+  private int rcvBuf;
+  private int linger;
+  private long buffer;
+  private String compression;
+  private final AtomicInteger rrc = new AtomicInteger(0);
+  private final Map<Properties, AdminClientCreateFunctionImpl> adminClientCreateFuncCache =
+      new ConcurrentHashMap<>();
+  private final Map<String, AdminClient> adminClientCache = new ConcurrentHashMap<>();
```
Comment (evtimo) [inline src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java:50]: As I understood, AdminClient doesn't need caching, it can be like final variable in constructor.

File: src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java (hunk 1 above)
Comment (AnastasiaShemyakinskaya) [reply src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java]: It's my mistake, I will fix it, thanks.

File: src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java (hunk 2 above)
Comment (akurilov) [reply src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java:50]: @zhenytim AdminClient depends on the selected Kafka node address, so it should be cached.

File: src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java [hunk 3]
```diff
@@ -12,13 +14,45 @@
...
+import org.apache.kafka.clients.admin.AdminClientConfig;
+import org.apache.kafka.clients.admin.NewTopic;
+import org.apache.kafka.clients.producer.*;
 
 public class KafkaStorageDriver<I extends Item, O extends Operation<I>>
     extends CoopStorageDriverBase<I, O> {
 
+  private final String[] endpointAddrs;
+  private final int nodePort;
+  private boolean key;
```
Comment (akurilov) [inline src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java]: Please use final for the fields initialized in the constructor

File: src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java [hunk 4]
```diff
@@ -27,17 +61,34 @@ public KafkaStorageDriver(
...
+    this.linger = netConfig.intVal("linger");
+    this.numPartition = 1;
+    this.replicationFactor = 1;
   }
 
   @Override
   protected final boolean submit(final O op) throws IllegalStateException {
     if (concurrencyThrottle.tryAcquire()) {
       final var opType = op.type();
+      var nodeAddr = op.nodeAddr();
```
Comment (akurilov) [inline src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java]: final

File: src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java [hunk 5]
```diff
@@ -71,7 +122,88 @@ private void submitRecordDeleteOperation() {}
 
   private void submitRecordReadOperation() {}
 
-  private void submitRecordCreateOperation() {}
+  boolean completeFailedOperation(final O op, final Throwable thrown) {
+    thrown.printStackTrace();
```
Comment (akurilov) [inline src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java]: Please use the logger instead:
`LogUtil.exception(Level.DEBUG, e, "{}: operation failed: {}", stepId, op);`

File: src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java [hunk 6]
```diff
@@ -71,7 +122,88 @@ private void submitRecordDeleteOperation() {}
...
+          adminClientCreateFuncCache.computeIfAbsent(
+              createAdminConfig(nodeAddr), AdminClientCreateFunctionImpl::new);
+      var adminClient = adminClientCache.computeIfAbsent(nodeAddr, adminConfig);
+      var producerConfig =
+          producerCreateFuncCache.computeIfAbsent(
+              createProducerConfig(nodeAddr), ProducerCreateFunctionImpl::new);
+      var kafkaProducer = producerCache.computeIfAbsent(nodeAddr, producerConfig);
+      var recordItem = recordOp.item();
+      var topicName = recordItem.name();
+      if (topicCache.get(topicName) == null) {
```
Comment (akurilov) [inline src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java]: The approach is not thread safe. Please use the method `computeIfAbsent` of `ConcurrentMap` interface to check and instantiate (if necessary) atomically. Also use some method reference instead of the constructor

File: src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java [hunk 7]
```diff
@@ -71,7 +122,88 @@ private void submitRecordDeleteOperation() {}
...
+        final NewTopic topic = new NewTopic(topicName, numPartition, replicationFactor);
+        topicCache.put(topicName, topic);
+        adminClient.createTopics(Collections.singletonList(topic));
+      }
+      Callback callback =
```
Comment (akurilov) [inline src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java]: Please use the method reference instead of callback. The driver class can implement the Callback interface for this either the callback may be a field of the driver. This is required because we don't want to create new callback instance every time.

File: src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java [hunk 8]
```diff
@@ -71,7 +122,88 @@ private void submitRecordDeleteOperation() {}
...
+          (RecordMetadata metadata, Exception exception) -> {
+            if (exception != null) {
+              completeFailedOperation((O) recordOp, exception);
+            } else {
+              completeOperation((O) recordItem, SUCC);
+            }
+          };
+      if (key) {
+        var producerKey = Long.toString(recordItem.offset(), Character.MAX_RADIX);
```
Comment (akurilov) [inline src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java]: Let's use item.name() - it would be much more simple

File: src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java [hunk 9]
```diff
@@ -12,13 +14,45 @@
...
 
 public class KafkaStorageDriver<I extends Item, O extends Operation<I>>
     extends CoopStorageDriverBase<I, O> {
 
+  private final String[] endpointAddrs;
+  private final int nodePort;
+  private boolean key;
+  private int numPartition;
+  private short replicationFactor;
+  private int request;
```
Comment (akurilov) [inline src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java]: rename to something like "requestSizeLimit"

File: src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java [hunk 10]
```diff
@@ -71,7 +122,88 @@ private void submitRecordDeleteOperation() {}
...
+        "org.apache.kafka.common.serialization.StringSerializer");
+    producerConfig.put(
+        ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG,
+        "com.emc.mongoose.storage.driver.kafka.io.DataItemSerializer");
+    return producerConfig;
+  }
+
+  private void submitRecordCreateOperation(final DataOperation recordOp, String nodeAddr) {
+    try {
+      var adminConfig =
```
Comment (akurilov) [inline src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java]: 1. Let's use the config cache in order to not create the config on each invocation
2. You can use single producer config instead of admin config too
3. Use `val` instead of `var` for the immutable values. This brings in a bit more safety

**Review by akurilov — CHANGES_REQUESTED** [review]

File: src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java [hunk 11]
```diff
@@ -75,7 +127,85 @@ private void submitRecordDeleteOperation() {}
...
+    try {
+      val config = configCache.computeIfAbsent(nodeAddr, this::createConfig);
+      val adminConfig =
+          adminClientCreateFuncCache.computeIfAbsent(config, AdminClientCreateFunctionImpl::new);
+      val adminClient = adminClientCache.computeIfAbsent(nodeAddr, adminConfig);
+      val producerConfig =
+          producerCreateFuncCache.computeIfAbsent(config, ProducerCreateFunctionImpl::new);
+      val kafkaProducer = producerCache.computeIfAbsent(nodeAddr, producerConfig);
+      val recordItem = recordOp.item();
+      val topicName = recordItem.name();
```
Comment (akurilov) [inline src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java]: The record item name will look like `/<somepath>/<somename>`. I suggest to use the `<somepath>` for the topic name and `<somename>` for the key. Example:
https://github.com/emc-mongoose/mongoose-storage-driver-pravega/blob/1f693952f82dfe86ff52f5bea7587d8816c46902/src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java#L467

**Review by akurilov — CHANGES_REQUESTED** [review]

File: src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java (hunk 11 above)
Comment (AnastasiaShemyakinskaya) [reply src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java]: As I understand topicName = recordOp.dstPath(), key = recordItem,name()

Comment (akurilov) [reply src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java]: Yes, if dstPath doesn't include item name (I'm unsure)

**Review by akurilov — DISMISSED** [review]

**Review by akurilov — APPROVED** [review]
